src/jit/lower.cpp

```diff
@@ -571,6 +579,268 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+    //   |  |  {  |  |  +--*  lclVar    int    V23 rat1
+    //   |  |  {  |  +--*  gt_long   long
+    //   |  |  {  \--*  storeIndir long
+    //   |  +--*  lclVar    ref    V11 tmp9         u:3 (last use) $21c
+    //   |  +--*  lclVar    ref    V02 tmp0         u:3 $280
+    //   |  +--*  const     int    8 $4a
+    //   \--*  call help void   HELPER.CORINFO_HELP_ARRADDR_ST $205
+    //  <linkEndNode>
+
+    GenTree* linkBegin = comp->fgGetFirstNode(tree)->gtPrev;
+    GenTree* linkEnd = tree->gtNext;
+    GenTree* gtLong = tree->gtOp.gtOp2;
+
+    // Save address to a temp. It is used in storeIndLow and storeIndHigh trees.
+    GenTreeStmt* addrStmt = comp->fgInsertEmbeddedFormTemp(&tree->gtOp.gtOp1);
+#ifdef DEBUG
+    if (comp->verbose)
+    {
+        printf("[DecomposeStoreInd]: Saving address tree to a temp var:\n");
+        comp->gtDispTree(addrStmt);
+    }
+#endif
+
+    // If we have made a new top-level statement, and it has inherited any
+    // embedded statements from curStmt, they have not yet been decomposed.
+    if (addrStmt->gtStmtIsTopLevel())
+    {
+        for (GenTreePtr nextEmbeddedStmt = addrStmt->gtStmtNextIfEmbedded();
+             nextEmbeddedStmt != nullptr;
+             nextEmbeddedStmt = nextEmbeddedStmt->gtStmt.gtStmtNextIfEmbedded())
+        {
+            comp->compCurStmt = nextEmbeddedStmt;
+            comp->fgWalkTreePost(&nextEmbeddedStmt->gtStmt.gtStmtExpr, &Lowering::DecompNodeHelper, this, true);
+        }
+    }
+
+    // Restore curStmt.
+    comp->compCurStmt = currStmt;
+
+    if (!gtLong->gtOp.gtOp1->OperIsLeaf())
+    {
+        GenTreeStmt* dataLowStmt = comp->fgInsertEmbeddedFormTemp(&gtLong->gtOp.gtOp1);
+#ifdef DEBUG
+        if (comp->verbose)
+        {
+            printf("[DecomposeStoreInd]: Saving low data tree to a temp var:\n");
+            comp->gtDispTree(dataLowStmt);
+        }
+#endif
+        // If we have made a new top-level statement, and it has inherited any
+        // embedded statements from curStmt, they have not yet been decomposed.
+        if (dataLowStmt->gtStmtIsTopLevel())
+        {
+            for (GenTreePtr nextEmbeddedStmt = dataLowStmt->gtStmtNextIfEmbedded();
+                 nextEmbeddedStmt != nullptr;
+                 nextEmbeddedStmt = nextEmbeddedStmt->gtStmt.gtStmtNextIfEmbedded())
+            {
+                comp->compCurStmt = nextEmbeddedStmt;
+                comp->fgWalkTreePost(&nextEmbeddedStmt->gtStmt.gtStmtExpr, &Lowering::DecompNodeHelper, this, true);
+            }
+        }
+
+        // Restore curStmt.
+        comp->compCurStmt = currStmt;
+    }
+
+    if (!gtLong->gtOp.gtOp2->OperIsLeaf())
+    {
+        GenTreeStmt* dataHighStmt = comp->fgInsertEmbeddedFormTemp(&gtLong->gtOp.gtOp2);
+#ifdef DEBUG
+        if (comp->verbose)
+        {
+            printf("[DecomposeStoreInd]: Saving high data tree to a temp var:\n");
+            comp->gtDispTree(dataHighStmt);
+        }
+#endif
+    }
```

Comment:
Right, without `GTF_REVERSE_OPS` `dataHighStmt` should never become a top-level statement, the `GT_LCL_VAR` node produced by `dataLowStmt` prevents that. I don't see any code setting `GTF_REVERSE_OPS` on `GT_LONG` nodes and hopefully no such code will ever be added as it would simply complicate things more. So not applying the workaround here seems the right thing to do.